### ISSUE TITLE
Lint to ban impl Trait in parameters

### ISSUE THREAD

**Issue by SUPERCILEX** [rust-lang/rust-clippy#10030]
### Advantage

You can't use the turbofish with parameter impls which makes it strictly less powerful as far as I can tell. This lint should only apply to public items.

Name: `impl_trait_in_param_of_pub_fn`.

### Category

style (or potentially pedantic)

### Example

```rust
fn foo(t: impl Trait) {}
```

Could be written as:

```rust
fn foo<T: Trait>(t: T) {}
```

**Comment by blyxyas** [comment]
Just to be clear, when you refer to `you can't use the turbofish with parameter impl`, can you put an example?

**Comment by SUPERCILEX** [comment]
Sure:
```rust
trait A {}

impl A for u32 {}
impl A for u64 {}

fn foo_impl(_a: impl A) {}
fn foo_turbo<AA: A>(_a: AA) {}

fn do_stuff() {
    // foo_impl(0); // Doesn't compile
    // foo_impl::<u32>(0); // Doesn't compile
    foo_impl(0 as u32);
    foo_turbo(0 as u32);
    foo_turbo::<u32>(0);
}
```

It gets more annoying when people use `Option<impl Something>` because then when you want to pass in None you have to say `None::<Blah>` which looks weird IMO. I guess those are both turbofish, so to be more precise I mean turbofish in the function generic position.

**Comment by SUPERCILEX** [comment]
Woohoo, thanks!

PS: https://github.com/rust-lang/rust-clippy/commit/5ef3cc8ca1af8df6421adf2e600b774c8e055589#r101419032
